pkg/engine/shield/absorb.go

```diff
@@ -15,21 +15,26 @@ func (mgr *Manager) AbsorbDamage(target key.TargetID, damage float64) float64 {
 	// 6. remaining state should be only shields that starting HP > damage
 
 	// placeholder just for some basic event emission
-	for id, shield := range mgr.targets[target] {
+	for _, shield := range mgr.targets[target] {
 		prevHP := shield.HP
 		shield.HP -= damage
 		if shield.HP < 0 {
 			shield.HP = 0
 		}
 		mgr.event.ShieldChange.Emit(event.ShieldChange{
-			ID:     id,
+			ID:     shield.name,
 			Target: target,
 			OldHP:  prevHP,
 			NewHP:  shield.HP,
 		})
 
 		if shield.HP == 0 {
-			mgr.RemoveShield(id, target)
+			mgr.RemoveShield(shield.name, target)
```

Comment:
You cannot remove entries from a list while iterating over a list. This will cause the iteration to break and incorrectly iterate over the list. See how modifier remove defers all the removal calls to after the iteration is complete (and builds a separate list of entries to remove):

https://github.com/simimpact/srsim/blob/12711171278f35df30a278955b67435af4708556/pkg/engine/modifier/remove.go#L9-L22